arch/risc-v/src/mpfs/mpfs_gpio.c

```diff
@@ -202,3 +318,145 @@ bool mpfs_gpioread(gpio_pinset_t pinset)
   return ((getreg32(baseaddr + MPFS_GPIO_GPIN_OFFSET) & (1 << pin)) ? 1 : 0);
 }
 
+/****************************************************************************
+ * Name: mpfs_gpiosetevent
+ *
+ * Description:
+ *   Sets/clears GPIO based event and interrupt triggers.
+ *
+ ****************************************************************************/
+
+int mpfs_gpiosetevent(gpio_pinset_t pinset, bool risingedge,
+                      bool fallingedge, bool high, bool low, bool event,
+                      xcpt_t func, void *arg)
+{
+  uint8_t pin = (pinset & GPIO_PIN_MASK) >> GPIO_PIN_SHIFT;
+  uint8_t bank = (pinset & GPIO_BANK_MASK) >> GPIO_BANK_SHIFT;
+  bool edge = risingedge || fallingedge;
+  bool level = high || low;
+
+  uint16_t irq_n = bank == 1 ? pin + GPIO_BANK0_NUM_PINS : pin;
+  uint16_t plic_irq;
+  struct gpio_callback_s *cb;
+  int ret;
+
+  if (bank > 2        ||   /* invalid bank */
+      (edge && level) ||   /* edge and level irq at the same */
+      (high && low)   ||   /* level high and low at the same */
+      irq_n >= GPIO_BANK0_NUM_PINS + GPIO_BANK1_NUM_PINS ||
+      (bank == 2 && irq_n >= GPIO_BANK2_NUM_PINS))
+    {
+      return -EINVAL;
+    }
+
+  /* if func is NULL, make sure to disable interrupt */
+
+  if (func == NULL)
+    {
+      risingedge = false;
+      fallingedge = false;
+      high = false;
+      low = false;
+    }
+
+  cb = bank == 2 ? &g_fab_gpio_callbacks[irq_n]
+                 : &g_mss_gpio_callbacks[irq_n];
+
+  /* make sure the GPIO is input */
+
+  pinset &= ~GPIO_MODE_MASK;
+  pinset |= GPIO_INPUT;
+
+  /* set the irq mask for the gpio */
+
+  pinset &= ~GPIO_IRQ_MASK;
+
+  if (edge)
+    {
+      if (risingedge && fallingedge)
+        {
+          pinset |= GPIO_IRQ_EDGE_BOTH;
+        }
+      else if (fallingedge)
+        {
+          pinset |= GPIO_IRQ_EDGE_NEG;
+        }
+      else if (risingedge)
+        {
+          pinset |= GPIO_IRQ_EDGE_POS;
+        }
+    }
+  else
+    {
+      if (high)
+        {
+          pinset |= GPIO_IRQ_HIGH;
+        }
+      else if (low)
+        {
+          pinset |= GPIO_IRQ_LOW;
+        }
+    }
+
+  /* if event, set ENABLE bit, otherwise leave it 0 */
+
+  if (event)
+    {
+      pinset |= GPIO_IRQ_ENABLE;
+    }
+
+  if (mpfs_configgpio(pinset) != OK)
+    {
+      return -EINVAL;
+    }
+
+  /* set/reset the GPIO_INTERRUPT_FAB_CR for bank 2 */
+
+  if (bank == 2)
+    {
+      if (event)
+        {
+          modifyreg32(MPFS_SYSREG_BASE +
+                      MPFS_SYSREG_GPIO_INTERRUPT_FAB_CR_OFFSET,
+                      0, SYSREG_GPIO_INTERRUPT_FAB_CR(pin));
+        }
+      else
+        {
+          modifyreg32(MPFS_SYSREG_BASE +
+                      MPFS_SYSREG_GPIO_INTERRUPT_FAB_CR_OFFSET,
+                      SYSREG_GPIO_INTERRUPT_FAB_CR(pin), 0);
+        }
+    }
+
+  /* calculate the plic vector location from gpio irq number */
+
+  plic_irq = irq_n + MPFS_IRQ_GPIO02_BIT0;
```

Comment:
I assume the same PLIC IRQ source is used for S-/M-mode ?

Reply:
yes; this is actually just an index. even if the register set start address would be different, that would be in irq_attach() / up_enable_irq() code